File: dp/maxSubArray.js
// 最大子串和
// 给定数组为 A = [1， 3， -2， 4， -5]， 则最大连续子序列和为 6，即 1 + 3 +（-2）+ 4 = 6
function maxSubArray(nums) {
    const dp = [];
    let max = nums[ 0 ];
    dp[ 0 ] = nums[ 0 ];
    for (let i = 1; i < nums.length; i++) {
        dp[ i ] = Math.max(nums[ i ], dp[ i - 1] + nums[ i ]);
        max = Math.max(max, dp[i]);
    }
    return max;
}

// dp[i]表示以i结尾的最大和
// dp[i] = Max(nums[i], d[i-1] + nums[i])
const nums = [ 1, 3, -2, 4, -5 ];
const ret = maxSubArray(nums);
console.log(ret);

File: dp/maxSubArray.go
package main

// 给你一个整数数组 nums ，请你找出一个具有最大和的连续子数组（子数组最少包含一个元素），返回其最大和。
// 子数组 是数组中的一个连续部分

func maxSubArray(nums []int) int {
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	ans := nums[0]
	for i := 1; i < len(nums); i++ {
		dp[i] = max(dp[i-1]+nums[i], nums[i])
		ans = max(ans, dp[i])
	}
	return ans
}

func main() {
	nums := []int{5, 4, -1, 7, 8}
	maxSubArray(nums)
}
